src/math/bvh.js

```diff
@@ -0,0 +1,456 @@
+            }
+            if (boundsMin === boundsMax) continue;
+
+            const bin = Array.apply(null, Array(this.bins)).map(function () {
+                return new Bin();
+            });
+            let scale = this.bins / (boundsMax - boundsMin);
+            for (let i = 0; i < node.triCount; i++) {
+                const triangle = triangles[triIdx[node.leftFirst + i]];
+                const binIdx = Math.min(this.bins - 1, Math.floor((triangle.centroid[b] - boundsMin) * scale));
+                bin[binIdx].triCount++;
+                bin[binIdx].bounds.setMinMax(vec3Min(bin[binIdx].bounds.getMin(), triangle.vertex0), vec3Max(bin[binIdx].bounds.getMax(), triangle.vertex0));
+                bin[binIdx].bounds.setMinMax(vec3Min(bin[binIdx].bounds.getMin(), triangle.vertex1), vec3Max(bin[binIdx].bounds.getMax(), triangle.vertex1));
+                bin[binIdx].bounds.setMinMax(vec3Min(bin[binIdx].bounds.getMin(), triangle.vertex2), vec3Max(bin[binIdx].bounds.getMax(), triangle.vertex2));
+            }
+            const leftArea = Array(this.bins - 1);
+            const rightArea = Array(this.bins - 1);
+            const leftCount = Array(this.bins - 1);
+            const rightCount = Array(this.bins - 1);
+
+            const leftBox = new BoundingBox(new Vec3(), new Vec3());
+            const rightBox = new BoundingBox(new Vec3(), new Vec3());
+            let leftSum = 0;
+            let rightSum = 0;
+
+            for (let i = 0; i < this.bins - 1; i++) {
+                leftSum += bin[i].triCount;
+                leftCount[i]  = leftSum;
+                leftBox.setMinMax(vec3Min(leftBox.getMin(), bin[i].bounds.getMin()), vec3Max(leftBox.getMax(), bin[i].bounds.getMax()));
+                leftArea[i] = leftBox.area();
+                rightSum += bin[this.bins - 1 - i].triCount;
+                rightCount[this.bins - 2 - i] = rightSum;
+                rightBox.setMinMax(vec3Min(rightBox.getMin(), bin[this.bins - 1 - i].bounds.getMin()), vec3Max(rightBox.getMax(), bin[this.bins - 1 - i].bounds.getMax()));
+                rightArea[this.bins - 2 - i] = rightBox.area();
+            }
+
+            scale = (boundsMax - boundsMin) / this.bins;
+            for (let i = 0; i < this.bins - 1; i++) {
+                const planeCost = leftCount[i] * leftArea[i] + rightCount[i] * rightArea[i];
+                if (planeCost < bestCost) {
+                    splitDetails.axis = b;
+                    splitDetails.splitPos = boundsMin + scale * (i + 1);
+                    bestCost = planeCost;
+                }
+            }
+        }
+        return bestCost;
+    }
+
+    /**
+     * Calculates the estimated cost of raycasting through a node
+     *
+     * @param {BVHNode} node - The node to be evaluated
+     * @returns {number} The cost of the raycast
+     * @ignore
+     */
+    calculateNodeCost(node) {
+        const extent = new Vec3();
```

Comment:
use some global variable to avoid the allocation each time here

Reply:
Added